Partial normalization of Lambda request and initial PR for migration to Envelop

dotansimha opened the pull request:
Hi, people of the Redwood!  :) 

I created an initial PR for migrating from apollo-server-lambda to Envelop and GraphQL-Helix. The goal of this PR is to normalize the incoming HTTP requests and try to handle them in a generic way. Also, since the request is detached from the handler, we can use any GraphQL library for execution. 

While `graphql-helix` provides the basic pipeline and the initial request normalization, `envelop` provides the connection to the GraphQL execution, and allow to enrich the entire GraphQL execution pipeline with custom code (custom context building, parser cache, validation cache, tracing, metrics collection and more). 

Changes made in this PR related to GraphQL execution:
- Migrate custom user context to be a pure function and run as Envelop plugin.
- Migrate auth context to be a pure function and run as Envelop plugin.
- Use GraphQL request pipeline provided by GraphQL-Helix
- Added GraphQL parser cache 
- Added GraphQL validation cache
- Allow users to provide custom Envelop plugins
- Use the latest GraphQL Playground directly from the source

Also in this PR:
- Move CORS logic to be standalone and generic. I guess this could be done with an external package as well.
- Move Healthcheck logic to be standalone and (almost) generic. I kept the original Apollo endpoint for backward compatibility.
- Use `Request` from `graphql-helix` in order to normalize some aspects of the incoming HTTP request. This could eventually allow the handler code to be clean and simple and not to depend on any specific Lambda provider. 

What's next? 
- Would love to get feedback! (@thedavidprice @peterp and anyone else of course)
- Testing and backward compatibility checks - I think you know best :D

dthyresson commented:
I'm still reading, learning about Envelop & `graphql-helix`but I am curious where one might be able to "plug in" our logger (it is added in each app with https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/lib/logger.ts) to optionally log GraphQL requests and errors.

Currently, we can import the logger into a service and log there, but one needs to do this for each service.

Might the be a way to provide the logger throughout?

dthyresson commented:
> Might the be a way to provide the logger throughout?

Oh I see an example like:

```js
const getEnveloped = envelop({
  plugins: [useSchema(schema), useLogger(), useTiming()],
});
```

Would that just be:

```js
const DEFAULT_OPTIONS: LoggerPluginOptions = {
  logFn: console.log,
};

export const useLogger = (rawOptions: LoggerPluginOptions = DEFAULT_OPTIONS): Plugin => {
  const options = {
    DEFAULT_OPTIONS,
    ...rawOptions,
  };
...
```

Setting `logFn` to `logger.debug`?

dotansimha commented:
> I'm still reading, learning about Envelop & `graphql-helix`but I am curious where one might be able to "plug in" our logger (it is added in each app with https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/lib/logger.ts) to optionally log GraphQL requests and errors.
> 
> Currently, we can import the logger into a service and log there, but one needs to do this for each service.
> 
> Might the be a way to provide the logger throughout?

Rebased and pushed :) 

Regarding the logger - if the goal is to bundle an existing logger, with a custom flow, you can do something like that: 

```ts
const useRedwoodLogger: Plugin = () => {
  return {
    onExecute({ args }) {
      logger.log('execution started...', { args });

      return {
        onExecuteDone: ({ result }) => {
           logger.log('execution ended...', { args, result });
        },
      };
    },
  }
}
```

I can add that to the PR - just wasn't sure how should I get the `logger` instance at the level of the handler? is it injected somewhere?

dthyresson commented:
> just wasn't sure how should I get the `logger` instance at the level of the handler? is it injected somewhere

The logger is created in the app via this https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/lib/logger.ts` which is added as as part of creating a new RW app. 

I imagine it would be added to the ``https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/functions/graphql.ts`` graphql function also created as part of a new app.

Perhaps import `logger` just like `db` is here:

https://github.com/redwoodjs/redwood/blob/76fb8141a04eba21034a0c5133f66aba2dc1197a/packages/create-redwood-app/template/api/src/functions/graphql.ts#L8

? 

Would that mean `useRedwoodLogger` would also be added as part of the `graphql` function?

dotansimha commented:
> > just wasn't sure how should I get the `logger` instance at the level of the handler? is it injected somewhere
> 
> The logger is created in the app via this [https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/lib/logger.ts`](https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/lib/logger.ts%60) which is added as as part of creating a new RW app.
> 
> I imagine it would be added to the `https://github.com/redwoodjs/redwood/blob/main/packages/create-redwood-app/template/api/src/functions/graphql.ts` graphql function also created as part of a new app.
> 
> Perhaps import `logger` just like `db` is here:
> 
> https://github.com/redwoodjs/redwood/blob/76fb8141a04eba21034a0c5133f66aba2dc1197a/packages/create-redwood-app/template/api/src/functions/graphql.ts#L8
> 
> ?
> 
> Would that mean `useRedwoodLogger` would also be added as part of the `graphql` function?

Sounds good. I added a commit with that. It takes the logger and passes it with the handler creator. Then it used to track execution start/end with some basic details. I guess this could be improved based on Redwood needs and tracing flow?

dthyresson commented on packages/api/src/functions/graphql.ts:
```diff
@@ -157,6 +162,44 @@ const useRedwoodGlobalContextSetter = (): Plugin<RedwoodGraphQLContext> => ({
+const useRedwoodLogger = (
+  baseLogger: BaseLogger
+): Plugin<RedwoodGraphQLContext> => {
+  const childLogger = baseLogger.child({ name: 'GraphQL ' })
```
👍  for use of the child logger

dthyresson commented:
I would be interested to see these plugins included by default:

* https://github.com/dotansimha/envelop/tree/main/packages/plugins/depth-limit

I know depth (and rate limiting) is something Redwood should have an answer for -- either in FW default or in docs how to setup as best GraphQL practices.

* https://github.com/dotansimha/envelop/tree/main/packages/plugins/generic-auth

I wonder if this might replace the `beforeResolver` approach here in PR: https://github.com/redwoodjs/redwood/pull/2272

* https://github.com/dotansimha/envelop/tree/main/packages/plugins/disable-introspection (Could this chamge     `introspection: isDevEnv` in https://github.com/redwoodjs/redwood/blob/3a65eee02730721047efe48a85d7445c93f5da58/packages/api/src/functions/graphql.ts#L106

peterp commented:
Just a heads up that I'm removing this code, but am getting some errors whilst trying to run it - I'll debug and report back

peterp commented:
*reviewing not removing, sorry.

dthyresson commented:
Running the e2e tasks raises the `ReferenceError: Headers is not defined` exception in `graphql.ts`:

```terminal
api | ReferenceError: Headers is not defined
api | 
api | 14    schema: makeMergedSchema({
api | 15      schemas,
api | 16      services: makeServices({ services }),
api | 17    }),
api | 18  
api | 19    onException: () => {
api | 20      // Disconnect from your database with an unhandled exception.
api | 21      db.$disconnect()
api | 22    },
api | 23  })
api | 24  
```

This stems from the `cors.ts` in

```js

export function createCorsContext(cors: CorsConfig | undefined) {
  // Taken from apollo-server-lambda
  const corsHeaders = new Headers()
```


To fix this, I have added

```js
import { Headers } from 'node-fetch'
```
Since looking at `apollo-server-lambda` it gets Headers from 

https://github.com/apollographql/apollo-server/blob/9267a79b974e397e87ad9ee408b65c46751e4565/packages/apollo-server-env/src/polyfills/fetch.js#L1

```js
export { default as fetch, Request, Response, Headers } from 'node-fetch';
```

See `apollo-server-lambda`:

https://github.com/apollographql/apollo-server/blob/9267a79b974e397e87ad9ee408b65c46751e4565/packages/apollo-server-lambda/src/ApolloServer.ts#L23

```js
import { Headers } from 'apollo-server-env';
```

Adding that no longer raises the error, however, when running my e2e tests, the cells failed as:

![image](https://user-images.githubusercontent.com/1051633/119668129-e40aed00-be04-11eb-9c04-25e4acbd21ac.png)

![image](https://user-images.githubusercontent.com/1051633/119668202-f6852680-be04-11eb-9977-8468ac78a2e3.png)

dthyresson commented:
I discovered that the handler wasn't able to extract the operationName, query and variables from the request because the body wasn't parsed.

```terminal
api | >>> request body {"operationName":"USERS","variables":{},"query":"query USERS {\n  users {\n    id\n    email\n    name\n    __typename\n  }\n}\n"}
api | >>> request method POST
api | >>> operationName undefined
api | >>> query undefined
api | >>> variables undefined
api | >>> processRequest
```

By parsing the body in `normalizeRequest`

```js
function normalizeRequest(event: APIGatewayProxyEvent): Request {
  return {
    headers: event.headers || {},
    method: event.httpMethod,
    query: event.queryStringParameters,
    body: event.body && JSON.parse(event.body),
  }
}
```

Now, the query, operationName are found:

```terminal
api | DEBUG [2021-05-26 15:48:01.311 +0000]: getGraphQLParameters
api |     requestBody: {
api |       "operationName": "USERS",
api |       "variables": {},
api |       "query": "query USERS {\n  users {\n    id\n    email\n    name\n    __typename\n  }\n}\n"
api |     }
api |     operationName: "USERS"
api |     query: "query USERS {\n  users {\n    id\n    email\n    name\n    __typename\n  }\n}\n"
api |     variables: {}
api | DEBUG [2021-05-26 15:48:01.311 +0000]: About to processRequest
api | INFO [2021-05-26 15:48:01.313 +0000] (GraphQL ): GraphQL execution started
api |     operationName: "USERS"
api | ERROR [2021-05-26 15:48:01.341 +0000] (GraphQL ): GraphQL execution completed with errors:
api |     operationName: "USERS"
api |     errors: [
api |       {
api |         "message": "Cannot return null for non-nullable field Query.users.",
api |         "locations": [
api |           {
api |             "line": 2,
api |             "column": 3
api |           }
api |         ],
api |         "path": [
api |           "users"
api |         ]
api |       }
api |     ]
api | 
api | Error: Cannot return null for non-nullable field Query.users.
api | 
```

But it as if the service to fetch users isn't found (I am logging a trace in the users service that doesn't appear to be reached). But still verifying that.

Checking in the JSON parse fix and some logging to help diagnose.

dthyresson commented:
@dotansimha Hi. I am struggling to figure out how to test and see perhaps why (my guess) is that perhaps the schema and resolver isn't mapped such that it cannot execute the query?

> api | Error: Cannot return null for non-nullable field Query.users.

Here is some sample schema for a basic User model

```js
model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
}
```

 with `users` as the service/resolver.

```terminal
api | >>>> schema
api | GraphQLSchema {
api |   __validationErrors: undefined,
api |   description: undefined,
api |   extensions: undefined,
api |   astNode: {
api |     kind: 'SchemaDefinition',
api |     description: undefined,
api |     directives: [],
api |     operationTypes: [ [Object], [Object] ],
api |     loc: { start: 555, end: 601 }
api |   },
api |   extensionASTNodes: [],
api |   _queryType: Query,
api |   _mutationType: Mutation,
api |   _subscriptionType: undefined,
api |   _directives: [ @include, @skip, @deprecated, @specifiedBy ],
api |   _typeMap: [Object: null prototype] {
api |     Date: Date,
api |     Time: Time,
api |     DateTime: DateTime,
api |     JSON: JSON,
api |     JSONObject: JSONObject,
api |     Redwood: Redwood,
api |     String: String,
api |     Query: Query,
api |     Int: Int,
api |     User: User,
api |     CreateUserInput: CreateUserInput,
api |     UpdateUserInput: UpdateUserInput,
api |     Mutation: Mutation,
api |     Boolean: Boolean,
api |     __Schema: __Schema,
api |     __Type: __Type,
api |     __TypeKind: __TypeKind,
api |     __Field: __Field,
api |     __InputValue: __InputValue,
api |     __EnumValue: __EnumValue,
api |     __Directive: __Directive,
api |     __DirectiveLocation: __DirectiveLocation
api |   },
api |   _subTypeMap: [Object: null prototype] {},
api |   _implementationsMap: [Object: null prototype] {}
api | }
api | >>>> services
api | {
api |   users_users: {
api |     users: [Function: users],
api |     user: [Function: user],
api |     createUser: [Function: createUser],
api |     updateUser: [Function: updateUser],
api |     deleteUser: [Function: deleteUser]
api |   }
api | }
api | >>>> resolvers
api | {
api |   Date: Date,
api |   Time: Time,
api |   DateTime: DateTime,
api |   JSON: JSON,
api |   JSONObject: JSONObject,
api |   Query: {
api |     redwood: [Function: redwood],
api |     users: [Function: users],
api |     user: [Function: user]
api |   },
api |   Mutation: {
api |     createUser: [Function: createUser],
api |     updateUser: [Function: updateUser],
api |     deleteUser: [Function: deleteUser]
api |   }
api | }
api | >>>> schemaOptions
api | undefined
api | >>>> resolverValidationOptions
api | undefined
api | >>>> inheritResolversFromInterfaces
api | undefined
api | >>>> after addResolversToSchema
api | GraphQLSchema {
api |   __validationErrors: undefined,
api |   description: undefined,
api |   extensions: undefined,
api |   astNode: {
api |     kind: 'SchemaDefinition',
api |     description: undefined,
api |     directives: [],
api |     operationTypes: [ [Object], [Object] ],
api |     loc: { start: 555, end: 601 }
api |   },
api |   extensionASTNodes: [],
api |   _queryType: Query,
api |   _mutationType: Mutation,
api |   _subscriptionType: undefined,
api |   _directives: [ @include, @skip, @deprecated, @specifiedBy ],
api |   _typeMap: [Object: null prototype] {
api |     Date: Date,
api |     Time: Time,
api |     DateTime: DateTime,
api |     JSON: JSON,
api |     JSONObject: JSONObject,
api |     Redwood: Redwood,
api |     String: String,
api |     Query: Query,
api |     Int: Int,
api |     User: User,
api |     CreateUserInput: CreateUserInput,
api |     UpdateUserInput: UpdateUserInput,
api |     Mutation: Mutation,
api |     Boolean: Boolean,
api |     __Schema: __Schema,
api |     __Type: __Type,
api |     __TypeKind: __TypeKind,
api |     __Field: __Field,
api |     __InputValue: __InputValue,
api |     __EnumValue: __EnumValue,
api |     __Directive: __Directive,
api |     __DirectiveLocation: __DirectiveLocation
api |   },
api |   _subTypeMap: [Object: null prototype] {},
api |   _implementationsMap: [Object: null prototype] {}
api | }
api | Done. Took 2922ms.
api | 
api | Error: Cannot return null for non-nullable field Query.users.
api | 
```

Much seems there, but for some reason I never see `users` actually called as I had logging in:

```js
export const users = () => {
  console.log('!!!!!!!!!!!!>>> IN users')
  return db.user.findMany()
}
```

But I never see that.

Any ideas where to look?

I started to read https://github.com/ardatan/graphql-tools/blob/dfd900577c9331952a4588191771d2dc2e5174a9/packages/schema/src/addResolversToSchema.ts

thinking maybe something isn't mapping?

```js
        Object.keys(resolverValue).forEach(fieldName => {
          if (!fieldName.startsWith('__')) {
            const fields = type.getFields();
            const field = fields[fieldName];

```

Or maybe

```js
schema = updateResolversInPlace
    ? addResolversToExistingSchema(schema, resolvers, defaultFieldResolver)
    : createNewSchemaWithResolvers(schema, resolvers, defaultFieldResolver);

  if (requireResolversForResolveType || requireResolversForResolveType !== 'ignore') {
    checkForResolveTypeResolver(schema, requireResolversForResolveType);
  }

  return schema;
```

Isn't returning, but not sure how to debug this.

I'll try to update the makeMergedSchema tests to match the data I've logged and perhaps I'll learn more where to look.

dotansimha commented:
> @dotansimha Hi. I am struggling to figure out how to test and see perhaps why (my guess) is that perhaps the schema and resolver isn't mapped such that it cannot execute the query?
> 
> > api | Error: Cannot return null for non-nullable field Query.users.
> 
> Here is some sample schema for a basic User model
> 
> ```js
> model User {
>   id    Int     @id @default(autoincrement())
>   email String  @unique
>   name  String?
> }
> ```
> 
> with `users` as the service/resolver.
> 
> ```
> api | >>>> schema
> api | GraphQLSchema {
> api |   __validationErrors: undefined,
> api |   description: undefined,
> api |   extensions: undefined,
> api |   astNode: {
> api |     kind: 'SchemaDefinition',
> api |     description: undefined,
> api |     directives: [],
> api |     operationTypes: [ [Object], [Object] ],
> api |     loc: { start: 555, end: 601 }
> api |   },
> api |   extensionASTNodes: [],
> api |   _queryType: Query,
> api |   _mutationType: Mutation,
> api |   _subscriptionType: undefined,
> api |   _directives: [ @include, @skip, @deprecated, @specifiedBy ],
> api |   _typeMap: [Object: null prototype] {
> api |     Date: Date,
> api |     Time: Time,
> api |     DateTime: DateTime,
> api |     JSON: JSON,
> api |     JSONObject: JSONObject,
> api |     Redwood: Redwood,
> api |     String: String,
> api |     Query: Query,
> api |     Int: Int,
> api |     User: User,
> api |     CreateUserInput: CreateUserInput,
> api |     UpdateUserInput: UpdateUserInput,
> api |     Mutation: Mutation,
> api |     Boolean: Boolean,
> api |     __Schema: __Schema,
> api |     __Type: __Type,
> api |     __TypeKind: __TypeKind,
> api |     __Field: __Field,
> api |     __InputValue: __InputValue,
> api |     __EnumValue: __EnumValue,
> api |     __Directive: __Directive,
> api |     __DirectiveLocation: __DirectiveLocation
> api |   },
> api |   _subTypeMap: [Object: null prototype] {},
> api |   _implementationsMap: [Object: null prototype] {}
> api | }
> api | >>>> services
> api | {
> api |   users_users: {
> api |     users: [Function: users],
> api |     user: [Function: user],
> api |     createUser: [Function: createUser],
> api |     updateUser: [Function: updateUser],
> api |     deleteUser: [Function: deleteUser]
> api |   }
> api | }
> api | >>>> resolvers
> api | {
> api |   Date: Date,
> api |   Time: Time,
> api |   DateTime: DateTime,
> api |   JSON: JSON,
> api |   JSONObject: JSONObject,
> api |   Query: {
> api |     redwood: [Function: redwood],
> api |     users: [Function: users],
> api |     user: [Function: user]
> api |   },
> api |   Mutation: {
> api |     createUser: [Function: createUser],
> api |     updateUser: [Function: updateUser],
> api |     deleteUser: [Function: deleteUser]
> api |   }
> api | }
> api | >>>> schemaOptions
> api | undefined
> api | >>>> resolverValidationOptions
> api | undefined
> api | >>>> inheritResolversFromInterfaces
> api | undefined
> api | >>>> after addResolversToSchema
> api | GraphQLSchema {
> api |   __validationErrors: undefined,
> api |   description: undefined,
> api |   extensions: undefined,
> api |   astNode: {
> api |     kind: 'SchemaDefinition',
> api |     description: undefined,
> api |     directives: [],
> api |     operationTypes: [ [Object], [Object] ],
> api |     loc: { start: 555, end: 601 }
> api |   },
> api |   extensionASTNodes: [],
> api |   _queryType: Query,
> api |   _mutationType: Mutation,
> api |   _subscriptionType: undefined,
> api |   _directives: [ @include, @skip, @deprecated, @specifiedBy ],
> api |   _typeMap: [Object: null prototype] {
> api |     Date: Date,
> api |     Time: Time,
> api |     DateTime: DateTime,
> api |     JSON: JSON,
> api |     JSONObject: JSONObject,
> api |     Redwood: Redwood,
> api |     String: String,
> api |     Query: Query,
> api |     Int: Int,
> api |     User: User,
> api |     CreateUserInput: CreateUserInput,
> api |     UpdateUserInput: UpdateUserInput,
> api |     Mutation: Mutation,
> api |     Boolean: Boolean,
> api |     __Schema: __Schema,
> api |     __Type: __Type,
> api |     __TypeKind: __TypeKind,
> api |     __Field: __Field,
> api |     __InputValue: __InputValue,
> api |     __EnumValue: __EnumValue,
> api |     __Directive: __Directive,
> api |     __DirectiveLocation: __DirectiveLocation
> api |   },
> api |   _subTypeMap: [Object: null prototype] {},
> api |   _implementationsMap: [Object: null prototype] {}
> api | }
> api | Done. Took 2922ms.
> api | 
> api | Error: Cannot return null for non-nullable field Query.users.
> api | 
> ```
> 
> Much seems there, but for some reason I never see `users` actually called as I had logging in:
> 
> ```js
> export const users = () => {
>   console.log('!!!!!!!!!!!!>>> IN users')
>   return db.user.findMany()
> }
> ```
> 
> But I never see that.
> 
> Any ideas where to look?
> 
> I started to read https://github.com/ardatan/graphql-tools/blob/dfd900577c9331952a4588191771d2dc2e5174a9/packages/schema/src/addResolversToSchema.ts
> 
> thinking maybe something isn't mapping?
> 
> ```js
>         Object.keys(resolverValue).forEach(fieldName => {
>           if (!fieldName.startsWith('__')) {
>             const fields = type.getFields();
>             const field = fields[fieldName];
> ```
> 
> Or maybe
> 
> ```js
> schema = updateResolversInPlace
>     ? addResolversToExistingSchema(schema, resolvers, defaultFieldResolver)
>     : createNewSchemaWithResolvers(schema, resolvers, defaultFieldResolver);
> 
>   if (requireResolversForResolveType || requireResolversForResolveType !== 'ignore') {
>     checkForResolveTypeResolver(schema, requireResolversForResolveType);
>   }
> 
>   return schema;
> ```
> 
> Isn't returning, but not sure how to debug this.
> 
> I'll try to update the makeMergedSchema tests to match the data I've logged and perhaps I'll learn more where to look.

Fixed! :) Also squashes all, rebased and update all new dependencies to latest :)

dthyresson commented:
FYI when I yarn install I am currently getting the

```
[2/4] 🚚  Fetching packages...
error jest-regex-util@27.0.1: The engine "node" is incompatible with this module. Expected version "^10.13.0 || ^12.13.0 || ^14.15.0 || >=15.0.0". Got "14.7.0"
error Found incompatible module.
```

error. Am working through this.

dthyresson commented:
Getting closer:

```terminal
api | DEBUG [2021-06-02 14:22:06.267 +0000]: >>>> This must be used when self-hosting RedwoodJS.
api | DEBUG [2021-06-02 14:22:06.268 +0000]: >>>> handlerFn response.
api |     response: {}
api | INFO [2021-06-02 14:22:06.270 +0000] (GraphQL ): GraphQL execution started
api |     operationName: "USER_EXAMPLES"
api | POST /graphql 500 2.521 ms - -
api | 
api | TypeError: Cannot read property 'statusCode' of undefined
api | 
api | at /Users/dthyresson/redwood-tests/helix-1/redwood/dev-server/dist/requestHandlers/awsLambda.js(expressResponseForLambdaResult):62
api | 57    };
api | 58  };
api | 59  
api | 60  const expressResponseForLambdaResult = (expressResFn, lambdaResult) => {
api | 61    const {
api | 62      statusCode = 200,
api | 63      headers,
api | 64      body = ''
api | 65    } = lambdaResult;
api | 66  
api | 67    if (headers) {
api | 
api | 1 requestHandler
api |   /Users/dthyresson/redwood-tests/helix-1/redwood/dev-server/dist/requestHandlers/awsLambda.js:146
api | 
api | 2 lambdaHandler
api |   /Users/dthyresson/redwood-tests/helix-1/redwood/dev-server/dist/http.js:69
api | 
api | DEBUG [2021-06-02 14:22:06.284 +0000] (GraphQL ): GraphQL result
api |     operationName: "USER_EXAMPLES"
api |     GraphQLResult: {
api |       "data": {
api |         "userExamples": [
api |           {
api |             "id": 1,
api |             "email": "mark@example.com",
api |             "name": "mark",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 2,
api |             "email": "bob@example.com",
api |             "name": "bob",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 3,
api |             "email": "jackie@example.com",
api |             "name": "jackie",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 4,
api |             "email": "alice@example.com",
api |             "name": "alice",
api |             "__typename": "UserExample"
api |           }
api |         ]
api |       }
api |     }
api | INFO [2021-06-02 14:22:06.284 +0000] (GraphQL ): GraphQL execution completed
api |     operationName: "USER_EXAMPLES"
api |     result: {
api |       "data": {
api |         "userExamples": [
api |           {
api |             "id": 1,
api |             "email": "mark@example.com",
api |             "name": "mark",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 2,
api |             "email": "bob@example.com",
api |             "name": "bob",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 3,
api |             "email": "jackie@example.com",
api |             "name": "jackie",
api |             "__typename": "UserExample"
api |           },
api |           {
api |             "id": 4,
api |             "email": "alice@example.com",
api |             "name": "alice",
api |             "__typename": "UserExample"
api |           }
api |         ]
api |       }
api |     }
api | DEBUG [2021-06-02 14:22:06.285 +0000]: response from processRequest
api |     response: {
api |       "body": "{\"data\":{\"userExamples\":[{\"id\":1,\"email\":\"mark@example.com\",\"name\":\"mark\",\"__typename\":\"UserExample\"},{\"id\":2,\"email\":\"bob@example.com\",\"name\":\"bob\",\"__typename\":\"UserExample\"},{\"id\":3,\"email\":\"jackie@example.com\",\"name\":\"jackie\",\"__typename\":\"UserExample\"},{\"id\":4,\"email\":\"alice@example.com\",\"name\":\"alice\",\"__typename\":\"UserExample\"}]},\"extensions\":{\"tracing\":{\"version\":1,\"startTime\":\"2021-06-02T14:22:06.270Z\",\"endTime\":\"2021-06-02T14:22:06.285Z\",\"duration\":14636797,\"execution\":{\"resolvers\":[{\"path\":[\"userExamples\"],\"parentType\":\"Query\",\"fieldName\":\"userExamples\",\"returnType\":\"[UserExample!]!\",\"startOffset\":101415,\"duration\":12915906},{\"path\":[\"userExamples\",0,\"id\"],\"parentType\":\"UserExample\",\"fieldName\":\"id\",\"returnType\":\"Int!\",\"startOffset\":13103265,\"duration\":665811},{\"path\":[\"userExamples\",0,\"email\"],\"parentType\":\"UserExample\",\"fieldName\":\"email\",\"returnType\":\"String!\",\"startOffset\":13132145,\"duration\":642597},{\"path\":[\"userExamples\",0,\"name\"],\"parentType\":\"UserExample\",\"fieldName\":\"name\",\"returnType\":\"String\",\"startOffset\":13150306,\"duration\":627384},{\"path\":[\"userExamples\",1,\"id\"],\"parentType\":\"UserExample\",\"fieldName\":\"id\",\"returnType\":\"Int!\",\"startOffset\":13220733,\"duration\":563911},{\"path\":[\"userExamples\",1,\"email\"],\"parentType\":\"UserExample\",\"fieldName\":\"email\",\"returnType\":\"String!\",\"startOffset\":13236514,\"duration\":557233},{\"path\":[\"userExamples\",1,\"name\"],\"parentType\":\"UserExample\",\"fieldName\":\"name\",\"returnType\":\"String\",\"startOffset\":13253322,\"duration\":542914},{\"path\":[\"userExamples\",2,\"id\"],\"parentType\":\"UserExample\",\"fieldName\":\"id\",\"returnType\":\"Int!\",\"startOffset\":13303305,\"duration\":494658},{\"path\":[\"userExamples\",2,\"email\"],\"parentType\":\"UserExample\",\"fieldName\":\"email\",\"returnType\":\"String!\",\"startOffset\":13322091,\"duration\":477516},{\"path\":[\"userExamples\",2,\"name\"],\"parentType\":\"UserExample\",\"fieldName\":\"name\",\"returnType\":\"String\",\"startOffset\":13433841,\"duration\":367416},{\"path\":[\"userExamples\",3,\"id\"],\"parentType\":\"UserExample\",\"fieldName\":\"id\",\"returnType\":\"Int!\",\"startOffset\":13552144,\"duration\":251140},{\"path\":[\"userExamples\",3,\"email\"],\"parentType\":\"UserExample\",\"fieldName\":\"email\",\"returnType\":\"String!\",\"startOffset\":13578591,\"duration\":226466},{\"path\":[\"userExamples\",3,\"name\"],\"parentType\":\"UserExample\",\"fieldName\":\"name\",\"returnType\":\"String\",\"startOffset\":13594848,\"duration\":211989}]}}}}",
api |       "statusCode": 200,
api |       "headers": {}
api |     }
```

The changes I made were:

```js
  return async (
    event: APIGatewayProxyEvent,
    context: LambdaContext
  ): Promise<void> => {
```

`Promise<void> vs void` ... and then 

```
          const response = handlerFn(event, context)
          logger.debug({ response }, '>>>> handlerFn response.')
          Promise.resolve(response)
```

Something in 

```
///redwood/packages/dev-server/src/requestHandlers/awsLambda.ts

const expressResponseForLambdaResult = (
  expressResFn: Response,
  lambdaResult: APIGatewayProxyResult
) => {
  const { statusCode = 200, headers, body = '' } = lambdaResult
```

I thnk is expecting a callback

```
const handlerCallback =
    (expressResFn: Response) =>
    (error: Error, lambdaResult: APIGatewayProxyResult) => {
      if (error) {
        expressResponseForLambdaError(expressResFn, error)
        return
      }

      expressResponseForLambdaResult(expressResFn, lambdaResult)
    }
```

and `expressResponseForLambdaResult` isn't getting the response it expects.

Ideas?

thedavidprice commented:
Passing CI! Woop woop 🚀

thedavidprice commented:
merged base-branch changes, which are E2E CI fixes #2707

dthyresson commented:
@peterp and @dotansimha I confirmed the all e2e tests passed locally.

I also created a new RW app based on this branch and scaffolded a new "UserExample" (ie a user with an id, email, and name) and have confirmed that CRUD queries and mutations all also work.

I think we are very close now to merging and then cleaning up some log output and deciding on the middleware set needed to secure the graphql endpoint and such.

What should the next steps be? Merge and test more in canary with existing apps?

peterp commented:
@dthyresson I would like to make this run in parallel to `api`, so that we can evaluate it (in our own production apps), and let other's continue to use Apollo.

I would like to rename this package to `@redwoodjs/graphql-server` and bring back `@redwoodjs/api`.

 Part of the reason for this is because we've tightly coupled our request handling to aws-lambas, and it's now baked into this package as well as api. I would like to move away from that as some stage.

dotansimha commented:
> Part of the reason for this is because we've tightly coupled our request handling to aws-lambas, and it's now baked into this package as well as api. I would like to move away from that as some stage.

Totally makes sense. I also recommend moving the normalization part outside the GraphQL handler, so it can be done with any HTTP server, and just pass a normalized object to the handler. Also the response from `processRequest` can be returned as-is to the HTTP server, and then the HTTP server can decide how to respond to it (this way you can support regular GraphQL, but also streaming). 
I think this way Redwood will be able to run on any cloud provider, with any HTTP server, and even offer Subscriptions/real-time in some environments that supports itit.

dthyresson commented:
@peterp Let me me if any of the rework is something you'd like me to take on, or if you will move foraward with 

> I would like to rename this package to @redwoodjs/graphql-server and bring back @redwoodjs/api.

thedavidprice commented:
> Totally makes sense. I also recommend moving the normalization part outside the GraphQL handler, so it can be done with any HTTP server, and just pass a normalized object to the handler. Also, the response from processRequest can be returned as-is to the HTTP server, and then the HTTP server can decide how to respond to it (this way you can support regular GraphQL, but also streaming).

Oh, yeah!!! 🚀

> I think this way Redwood will be able to run on any cloud provider, with any HTTP server, and even offer Subscriptions/real-time in some environments that support it.

Also 🚀🚀

dthyresson commented:
Recommended plugins:

These plugins are the basic:

* useParserCache
* useValidationCache
* useErrorHandler (for printing errors with full stack traces to the log)
* useFilterAllowedOperations (in case you don’t have Subscriptions support)

And, in addition, in production:

* Maybe useGraphQLJit (for improved performance )
* tracing - with any one of the plugins (opentelemetry, sentry, new relic, datadog) 
* useMaskedErrors - to make sure we don’t expose sensitive error details
* useLogger - prints operation start/end with operation name, maybe variables (but after filtering it from sensitive data?), and result GraphQL errors .
* useDisableIntrospection

And, in addition, for development:

* useLogger that prints a lot of details
* useApolloTracing (for Playground tracing details)
* useRedwoodMockedAuth (doesn’t exists, but maybe just to make it easier for development? )

And some other plugins worth considering :

* usePreloadAssets - you can now return from the GraphQL metadata regarding assets, and * client-side can preload it.
* useExtendedValidation - for enriched validations that needs variables as well.
* useRateLimiter - probably needs user config.
* useDepthLimit - probably needs user config.

dthyresson commented:
@peterp has tasked me to take this current PR and get it ready such that people can elect to use Apollo Server (aka `api` aka classic RedwoodJS) or envelop (aka `graphql-server`).

This entails:

* having the code in this branch of `api` become the `graphql-server` package
* restoring current main `api` package

Then figuring out how to use one or the other.

jtoar commented:
@dthyresson with this PR, do you know if it'll be possible to set the GraphQL playground's endpoint from `redwood.toml`? #912

dthyresson commented:
> @dthyresson with this PR, do you know if it'll be possible to set the GraphQL playground's endpoint from `redwood.toml`? #912

Not sure. But I'll keep it in mind.

dthyresson commented:
> @peterp has tasked me to take this current PR and get it ready such that people can elect to use Apollo Server (aka `api` aka classic RedwoodJS) or envelop (aka `graphql-server`).
> 
> This entails:
> 
> * having the code in this branch of `api` become the `graphql-server` package
> * restoring current main `api` package
> 
> Then figuring out how to use one or the other.

@peterp I think I must have misunderstood but would not having `api` and also `graphql-server` make `graphql-server` another workspace?

Is this what you'd like?

dthyresson commented:
> @dthyresson with this PR, do you know if it'll be possible to set the GraphQL playground's endpoint from `redwood.toml`? #912

@jtoar I had a look here:

https://github.com/contrawork/graphql-helix/blob/c0122ac12c6158ec69260225363ec17a273257d1/packages/core/README.md#types

```js
export interface RenderGraphiQLOptions {
  /**
   * An optional GraphQL string to use when no query is provided and no stored
   * query exists from a previous session.  If undefined is provided, GraphiQL
   * will use its own default query.
   */
  defaultQuery?: string;
  /**
   * Whether to open the variable editor by default. Defaults to `true`.
   */
  defaultVariableEditorOpen?: boolean;
  /**
   * The endpoint requests should be sent. Defaults to `"/graphql"`.
   */
  endpoint?: string;
  /**
   * The initial headers to render inside the header editor. Defaults to `"{}"`.
   */
  headers?: string;
  /**
   * Whether the header editor is enabled. Defaults to `true`.
   */
  headerEditorEnabled?: boolean;
  /**
   * A cryptographic nonce for use with Content-Security-Policy.
   */
  nonce?: string;
  /**
   * The endpoint subscription requests should be sent to. Defaults to the value of the `endpoint` parameter.
   */
  subscriptionsEndpoint?: string;
}
```

And it looks like you can specify the endpoint that `renderGraphiQL()` would use.

```
 * The endpoint requests should be sent. Defaults to `"/graphql"`.
 * ```

So - yes - you *can* set the endpoint -- however, I do not think that the redwood.toml or it's config is available api side. So, that might be an issue. And of course, this is only in dev.

Tobbe commented:
> however, I do not think that the redwood.toml or it's config is available api side.

Surely it must me. It's where we configure the port to use for the api side, the path to the Prisma schema file etc

dthyresson commented:
> > however, I do not think that the redwood.toml or it's config is available api side.
> 
> Surely it must me. It's where we configure the port to use for the api side, the path to the Prisma schema file etc

Might be ... at build ... but I don't think at runtime since `redwood.toml` isn't deployed.

thedavidprice commented:
Next Steps for this PR are continuing here in #2827

@dthyresson please close this when you're ready. I've changed to draft for now.

dthyresson commented:
Per @thedavidprice am closing in favor of https://github.com/redwoodjs/redwood/pull/2827 which includes envelop v03.x, additional plugins, and api package separation.

Please follow that PR as we merge into main for an upcoming release.
